### PR TITLE
[BUGFIX] Fix interpolation on split geometry

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

Matthias, I think I found the problem with the interpolation on splitting geometries https://github.com/qgis/QGIS/issues/33489

QGIS uses GEOS algorithm to split geometries.If I'm not wrong,  in GEOS, using 3D point will returns an interpolation value which is the mean between geometries. On the contrary, in our logic, the interpolation is a linear interpolation on the split point.

If we drop Z (and M) value, we could have the "correct" interpolation.

image

### CODE REVIEW

**Comment by m-kuhn** [comment]
Looks good.
Can you add information about this interpolation to the docstring?

**Review by m-kuhn — APPROVED** [review]

File: src/core/geometry/qgsgeometry.cpp [hunk 1]
```diff
@@ -829,6 +829,15 @@ QgsGeometry::OperationResult QgsGeometry::splitGeometry( const QgsPointSequence
...
   QgsLineString splitLineString( splitLine );
+
+  /**
+   * QGIS uses GEOS algorithm to split geometries.
+   * Using 3D points in GEOS will returns an interpolation value which is the
+   * mean between geometries.
+   * On the contrary, in our logic, the interpolation is a linear interpolation
+   * on the split point. By dropping Z/M value, GEOS will returns the expected
+   * result. See https://github.com/qgis/QGIS/issues/33489
+   */
```
Comment (m-kuhn) [inline src/core/geometry/qgsgeometry.cpp:840]: Can you put the docstring to the header so it's visible in the api docs?
It shouldn't mention any implementation details like geos and mainly focus on the expected result. The one here is very nice as reference for future readers!

File: python/core/auto_generated/geometry/qgsgeometry.sip.in [hunk 2]
```diff
@@ -860,13 +860,22 @@ Splits this geometry according to a given line.
 Splits this geometry according to a given line.
 
 :param splitLine: the line that splits the geometry
-\param[out] newGeometries list of new geometries that have been created with the split
+\param[out] newGeometries list of new geometries that have been created with the ``splitLine``. If the geometry is 3D, a linear interpolation is performed on the geometry at split points, see Example.
```
Comment (m-kuhn) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in:863]:
```suggestion
\param[out] newGeometries list of new geometries that have been created with the ``splitLine``. If the geometry is 3D, a linear interpolation of the z value is performed on the geometry at split points, see example.
```

File: python/core/auto_generated/geometry/qgsgeometry.sip.in [hunk 3]
```diff
@@ -860,13 +860,22 @@ Splits this geometry according to a given line.
...
 :param splitFeature: Set to True if you want to split a feature, otherwise set to False to split parts
                      fix this bug?
 
 :return: OperationResult a result code: success or reason of failure
+
+* Example:
+.. code-block:: python
+
+      geometry = QgsGeometry.fromWkt('CompoundCurveZ ((2749546.2003820720128715 1262904.45356595050543547 100, 2749557.82053794478997588 1262920.05570670193992555 200))')
+      split_line = [QgsPoint(2749544.19, 1262914.79, 0), QgsPoint(2749557.64, 1262897.30, 0)]
```
Comment (m-kuhn) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in:875]:
```suggestion
      split_line = [QgsPoint(2749544.19, 1262914.79), QgsPoint(2749557.64, 1262897.30)]
```

Wouldn't that be more concise?

**Review by m-kuhn — APPROVED** [review]
Nice!

**Comment by nyalldawson** [comment]
Is this ready to merge? @m-kuhn ?

**Comment by m-kuhn** [comment]
Yes, thanks @lbartoletti !

**Comment by ponceta** [comment]
@lbartoletti @m-kuhn a backport to 3.10 would be possible? Or is it breaking stuff?

**Comment by m-kuhn** [comment]
I am in favor of a backport, @lbartoletti ?

**Comment by lbartoletti** [comment]
> I am in favor of a backport, @lbartoletti ?

Oh yes, of course! It's opened here https://github.com/qgis/QGIS/pull/36514#issuecomment-644006113

**Comment by ponceta** [comment]
Thanks!

**Comment by ponceta** [comment]
Tried on nightly today, works fine! Thanks a lot!

**Pull request merged**